### ISSUE TITLE
[TCK Challenge]: ee.jakarta.tck.data.web.validation.Rectangles has two bad methods

### ISSUE THREAD

**Issue by starksm64** [jakartaee/data#657]
### Specification

@Delete, spec/src/main/asciidoc/method-query.asciidoc

### Assertion

Automatic query methods, QBN findBy...

### TCK Version

SNAPSHOT (Locally built)

### Implementation being tested

HIbernate

### Challenge Scenario

Something else

### Full Description

The `ee.jakarta.tck.data.web.validation.Rectangles` repository interface has what looks to be two invalid methods:

```java
@Repository
public interface Rectangles extends DataRepository<Rectangle, String> {
...
    void deleteAllBy();

    @Size(min = 0, max = 3)
    List<Rectangle> findAllByOrderByIdAsc();
}
```

### Additional Context

The `deleteAllBy` method is not a QBN method, and its usage in the TCK implies it should be annotated with `@Delete` to function as an automatic query method that deletes all entities in the repo.

The `findAllByOrderByIdAsc` method fails to parse as a QBN method because there is no predicate between the By and OrderBy, and the current documented QBN grammar does not allow for an empty predicate. It does not seem like there is an approriate QBN method for the given TCK usage to attempt to return all entities and order them by id. It seems like the correct method would be:
```java
    @Size(min = 0, max = 3)
    @Find
    @OrderBy(value = "id")
    List<Rectangle> findAllByOrderByIdAsc();
```

### Is there an existing challenge for this?

- [X] I have searched the existing issues

**Comment by gavinking** [comment]
I wonder if this is just an error in the grammar. Perhaps it was supposed to be:

```
<query-method> ::= <subject> [<predicate>] [<order-clause>]
<subject> ::= (<action> | "find" <find-expression>) [<ignored-text>] "By"
<action> ::= "find" | "delete" | "update" | "count" | "exists"
```

I also wonder why the grammar isn't:

```
<query-method> ::= <subject> [<ignored-text>] <restriction> [<order-clause>]
<subject> ::= <action> | "find" <find-expression>
<restriction> ::= "All" | "By" <predicate>
<action> ::= "find" | "delete" | "update" | "count" | "exists"
```

Then sensible things like `findAllOrderByIdAsc()` and `deleteAll()` would be legal, but insensible things like `deleteAllBy()`, `countBy(), and `findAllBy()` would not.

**Comment by gavinking** [comment]
While I'm looking at the grammar, I also wonder why it's necessary to allow ugly things like `findByNameOrderByNameAscId`.

Wouldn't it be better to simplify this:

```
<order-clause> ::= "OrderBy" { <order-item> } ( <order-item> | <property> )
<order-item> ::= <property> ("Asc" | "Desc")
```

To this:

```
<order-clause> ::= "OrderBy" (<property> | <order-item> {<order-item>})
<order-item> ::= <property> ("Asc" | "Desc")
```

**Comment by gavinking** [comment]
> I also wonder why the grammar isn't:
> 
> ```
> <query-method> ::= <subject> [<ignored-text>] <restriction> [<order-clause>]
> <subject> ::= <action> | "find" <find-expression>
> <restriction> ::= "All" | "By" <predicate>
> <action> ::= "find" | "delete" | "update" | "count" | "exists"
> ```

Oh no, that doesn't work either, since it accepts `findFIrst10All()`.

**Comment by gavinking** [comment]
In https://github.com/jakartaee/data/pull/658 I propose to sort out these things.

**Comment by njr-11** [comment]
To fix the immediate issue with the TCK, I would say these methods should be switched to `@Delete` and `@Find` with `@OrderBy` as Scott suggests (and the methods could be given better names as well). That repository is under the validation package and I'm pretty sure its real intent is to cover interoperability with Jakarta Validation, not to cover Query by Method Name.  It was likely written before we had the grammar and before a lot of other things that went into the spec. 

This also brings up some issue with Query by Method Name with respect to finding/deleting all entities, which it looks like Gavin already has a solution in the above pull. We can use that pull for discussion.

**Comment by KyleAure** [comment]
Agreed, a lot of what we are finding now is that as the spec has been updated the TCK test have become invalid. 
I think we should make sure to work that calculation into our timeline, the more changes we make the the spec the more we break tests in the TCK.

I'd like to see TCK tests updated alongside changes to the spec both for consistancy and also so we can all see examples of how spec updates changes user applications.

I'll go ahead and fix these tests.

**Comment by gavinking** [comment]
I believe that this issue is fixed by #658.
